### PR TITLE
Remove some unnecessary unsafe usage

### PR DESCRIPTION
**Pull request by stephentoub**
This is mostly removing `unsafe` from methods that don't actually need it.

In a handful of places, I removed an `unsafe` by also removing the pointer-based code.  I'll call those out with comments.

In a few places, I moved the `unsafe` around, e.g. from being an unsafe block surrounding the entire method body to being in the signature.

And in a few places, I cleaned up some formatting thing right around the other code being modified.

cc: @GrabYourPitchforks 
Related to https://github.com/dotnet/roslyn/issues/48031

### CODE REVIEW

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.Stat.Span.cs [hunk 1]
```diff
@@ -28,10 +27,9 @@ internal static unsafe int Stat(ReadOnlySpan<char> path, out FileStatus output)
         [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_LStat", SetLastError = true)]
         internal static extern int LStat(ref byte path, out FileStatus output);
 
-        internal static unsafe int LStat(ReadOnlySpan<char> path, out FileStatus output)
+        internal static int LStat(ReadOnlySpan<char> path, out FileStatus output)
         {
-            byte* buffer = stackalloc byte[StackBufferSize];
-            var converter = new ValueUtf8Converter(new Span<byte>(buffer, StackBufferSize));
+            var converter = new ValueUtf8Converter(stackalloc byte[StackBufferSize]);
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.Stat.Span.cs:32]: These two functions were changed to use span-based stackalloc.

File: src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs [hunk 2]
```diff
@@ -14,10 +14,24 @@ internal static partial class IPv4AddressHelper
...
+
+                for (; (start < end) && (ch = str[start]) != '.' && ch != ':'; ++start)
+                {
+                    b = (b * 10) + ch - '0';
+                }
+
+                numbers[i] = (byte)b;
+                ++start;
+            }
+
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs:34]: Another place where I removed the unsafe code in favor of using spans.  Perf is unchanged, with the JIT eliminating the bounds checks on the numbers span.  This was the sole call site to ParseCanonical, so I just inlined it.

**Comment by GrabYourPitchforks** [comment]
Wonder if the p/invoke signatures or structs were generated straight from a .h file by a tool, hence the unnecessary unsafes?

**Comment by stephentoub** [comment]
> Wonder if the p/invoke signatures or structs were generated straight from a .h file by a tool, hence the unnecessary unsafes?

Maybe?  Don't know.  If they were produced by a tool at one point, they're not maintained that way now.

File: src/libraries/Common/src/Interop/FreeBSD/Interop.Process.cs [hunk 3]
```diff
@@ -86,10 +85,12 @@ internal static unsafe int[] ListAllPids()
...
             sysctlName[3] = pid;
 
-            int ret = Interop.Sys.Sysctl(sysctlName, ref pBuffer, ref bytesLength);
-            if (ret  != 0 ) {
+            if (Interop.Sys.Sysctl(sysctlName, ref pBuffer, ref bytesLength) != 0)
+            {
                 return null;
             }
+
+            // TODO Fix freeing of pBuffer: https://github.com/dotnet/runtime/issues/43418
```
Comment (GrabYourPitchforks) [inline src/libraries/Common/src/Interop/FreeBSD/Interop.Process.cs:93]: Heads up: possible merge conflict with https://github.com/dotnet/runtime/pull/43431, depending on checkin order.

File: src/libraries/Common/src/Interop/Windows/NCrypt/Interop.Properties.cs [hunk 4]
```diff
@@ -20,62 +20,43 @@ internal static partial class NCrypt
...
-                }
+                ErrorCode errorCode = Interop.NCrypt.NCryptGetProperty(
+                    hObject,
+                    pszProperty,
+                    pResult,
+                    sizeof(byte),
+                    out int cbResult,
+                    options);
+
+                Debug.Assert(errorCode != ErrorCode.ERROR_SUCCESS || cbResult == sizeof(byte));
```
Comment (GrabYourPitchforks) [inline src/libraries/Common/src/Interop/Windows/NCrypt/Interop.Properties.cs:35]: Nit: I find the earlier pattern (`if (success) { assert( /* ... */ ); }`) easier to read.

(Same comment applies elsewhere in this file.)

**Review by GrabYourPitchforks — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
My comments are just nits. Feel free to disregard at your discretion.

File: src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs [hunk 5]
```diff
@@ -14,10 +14,24 @@ internal static partial class IPv4AddressHelper
...
             return (numbers[0] << 24) + (numbers[1] << 16) + (numbers[2] << 8) + numbers[3];
```
Comment (GrabYourPitchforks) [inline src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs:35]: Can this be changed to `return BinaryPrimitives.ReadInt32BigEndian(numbers);`? I'm working on a separate PR for addressing this type of code pattern where I see it, but might be easier to address it in this PR so I don't introduce a conflict on my side.

**Review by GrabYourPitchforks — APPROVED** [review]

File: src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs:35]: Sure. Thanks.

File: src/libraries/Common/src/Interop/Windows/NCrypt/Interop.Properties.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/Common/src/Interop/Windows/NCrypt/Interop.Properties.cs:35]: Reverted

**Pull request merged**